### DISCUSSION TITLE
Error: no suitable C++ overload found in implementation of function lattice.lattice_dyson_g0_wk

### DISCUSSION THREAD

**Discussion by Lanlin123** [TRIQS/triqs#847] [Q&A] [answered]
Dear Triqs Developers,
     I am using triqs_tprf on singe-orbital kagome hubbard model by your example on the square lattice. But something went wrong. Below is the code:
    
   from triqs_tprf.tight_binding import TBLattice
  t = 0.1
H = TBLattice(
    units = [(6.83, 0, 0), (-3.42, 5.92, 0)],
    hopping = {
        # nearest neighbour hopping -t
        ( 0,0 ): [[0,t,t],
                  [t,0,t],
                  [t,t,0]],
                  
        ( 1,0 ): [[0,0,0],
                  [0,0,t],
                  [0,0,0]],
        
       ( -1,0 ): [[0,0,0],
                  [0,0,0],
                  [0,t,0]],
        
        ( 0,1 ): [[0,0,0],
                  [t,0,0],
                  [0,0,0]],
        
        ( 0,-1 ):[[0,t,0],
                  [0,0,0],
                  [0,0,0]],
        
        ( 1,1 ): [[0,0,0],
                  [0,0,0],
                  [0,0,0]],
        
      ( -1,-1 ): [[0,0,0],
                  [0,0,0],
                  [0,0,0]],
        
       ( -1,1 ): [[0,0,0],
                  [0,0,0],
                  [0,0,0]],
        
        ( 1,-1 ):[[0,0,0],
                  [0,0,0],
                  [0,0,0]],
        },
    orbital_positions = [(0.0568,0.5284,0),(0.4716,0.5284,0),(0.4716,0.9432,0)],
    orbital_names = ['1', '2','3'],
    )
nk = 32
kmesh=H.get_kmesh((nk, nk, 1))
e_k = H.dispersion(kmesh)
print(e_k)

from triqs.gf import MeshImFreq
from triqs_tprf.lattice import lattice_dyson_g0_wk
wmesh = MeshImFreq(beta=10, S='Fermion', n_max=32)
g0_wk = lattice_dyson_g0_wk(mu=0, e_k=e_k, mesh=wmesh)

TypeError                                 Traceback (most recent call last)
Input In [5], in <cell line: 5>()
      2 from triqs_tprf.lattice import lattice_dyson_g0_wk
      4 wmesh = MeshImFreq(beta=10, S='Fermion', n_max=32)
----> 5 g0_wk = lattice_dyson_g0_wk(mu=0, e_k=e_k, mesh=wmesh)

TypeError: Error: no suitable C++ overload found in implementation of function lattice.lattice_dyson_g0_wk

lattice_dyson_g0_wk(double mu, triqs_tprf::e_k_cvt e_k, gf_mesh<triqs::mesh::imfreq> mesh) -> triqs_tprf::g_wk_t 
 failed with the error : 
  Cpp2py converter: Python to C++ :
  ... Conversion of a Gf from Python to C++ triqs::gfs::gf_view<triqs::mesh::brzone, triqs::gfs::matrix_valued, nda::C_stride_layout, triqs::gfs::default_evaluator>
  ... Cannot convert the data of Gf from Python type :  numpy.ndarray to the C++ type nda::basic_array_view<std::complex<double>, 3, nda::C_stride_layout, (char)65, nda::default_accessor, nda::borrowed>

I wonder can you give me some advice. Thank you!

best,
Lily.

**Comment by the-hampel** [comment, marked as answer]
Hi Lily,
The problem here is that you used 
```
e_k = H.dispersion(kmesh)
```
to extract epsilon(k). This function will Fourier transform the hopping Hamiltonian AND diagonalize it and give pack only the eigenalues at each kpt. Hence: 
```
print(e_k.data.shape)
>> (1024, 3)
```
However, the function to construct the lattice Gf `lattice_dyson_g0_wk` expects an array at each kpoint of the orbital shape (3x3 here). The correct function to construct `e_k` is called just `fourier`:
```
e_k = H.fourier(kmesh)
print(e_k.data.shape)
>> (1024, 3, 3)
```
this function does also a FT but skips the diagonalization. 

I think the docstring of the `lattice_dyson_g0_wk` is a bit misleading as it asks for the dispersion. But the wording is just inconsistent with the latest triqs. We documented the functions fourier and dispersion on https://triqs.github.io/triqs/unstable/documentation/python_api/triqs.lattice.lattice_tools.TightBinding.html. Anyway changing `dispersion` to `fourier` solves your problem and the code should run. 

Best,
Alex

**Reply by Lanlin123** [reply]
Hi Alex,
    Thank you so much! It really works. Unluckily, there occurs another error that the plotting function plot_delta() and plot_chi() is not defined. Should I import any other module? I searched on the triqs website but couldn't find it . Below is the error:

NameError                                 Traceback (most recent call last)
Input In [19], in <cell line: 1>()
----> 1 plot_delta(deltas_freq_even_mom_even[0], lambdas_freq_even_mom_even[0])

NameError: name 'plot_delta' is not defined 

Best regards,
Lily.

**Reply by the-hampel** [reply]
I am not sure where these functions come from? Did you take these from https://triqs.github.io/tprf/latest/user_guide/Square%20lattice%20susceptibility.html? Can you point to the example you are trying to follow?

**Reply by Lanlin123** [reply]
Hi Alex,
    Thank you! I am following the example on solving the linearized Eliashberg equation in the random phase approximation limit from https://triqs.github.io/tprf/latest/user_guide/Solving%20the%20linearized%20Eliashberg%20equation%20in%20the%20random%20phase%20approximation%20limit.html.

**Reply by the-hampel** [reply]
This is indeed a "bug" in the tutorial. The function is not an official triqs or tprf function. You can find the function `plot_chi` in this tutorial: 
https://triqs.github.io/tprf/latest/user_guide/Bethe-Salpeter%20Equation%20on%20the%20Hubbard%20atom.html :
```
def plot_chi(chi, label='', opt={}):
    data = chi[Idx(1), :, :][0, 0, 0, 0].data

    plt.figure(figsize=(3.25*2, 3))
    subp = [1, 2, 1]

    def plot_data(data, subp, label=None, opt={}):
        plt.subplot(*subp); subp[-1] += 1
        plt.imshow( data, **opt )
        plt.title(label); plt.colorbar(); plt.axis('equal');

    plot_data(data.real, subp, label='Re[%s]' % label)
    plot_data(data.imag, subp, label='Im[%s]' % label)
    plt.tight_layout()
```
and the function `plot_delta` should look like this I think: 
```
def plot_delta(delta):
    import matplotlib.pyplot as plt

    fig, axes = plt.subplots(3, 3)

    vmax = np.max(np.abs(delta[Idx(0),:].data))

    for orb1, orb2 in itertools.product(list(range(p.norb)), repeat=2):
        shape = (p.nk, p.nk, p.norb, p.norb)
        data = delta[Idx(0), :].data.reshape(shape)
        plt.sca(axes[orb1,orb2])
        plt.imshow(data[:,:,orb1,orb2].real, cmap="RdBu_r",
                                                 vmax=vmax, vmin=-vmax)
        plt.colorbar()

    plt.sca(axes[-1,-1])
    for orb1, orb2 in itertools.product(list(range(p.norb)), repeat=2):
        plt.plot(delta.data[:, 1, orb1, orb2].real)
        plt.plot(delta.data[:, 1, orb1, orb2].imag)

    plt.show()
```
We need to correct the TPRF tutorials there. However, both are not free functions in triqs. There are only implemented for the purpose of the tutorial. You can try to use them but no guarantee. 

Best,
Alex

**Reply by Lanlin123** [reply]
Thank you so much, Alex.
